pkg/kudoctl/cmd/diagnostics/processing_context.go

```diff
@@ -0,0 +1,57 @@
+package diagnostics
+
+import (
+	"fmt"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1beta1"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// processingContext - shared data for the resource collectors
+// provides property accessors allowing to define a collector before the data it needs is available
+// provides update callback functions. callbacks panic if called on a wrong type of runtime.Object
+type processingContext struct {
```

Comment:
This context should be static as that would simplify resource collection a lot. Currently, it's being mutated by the collection runners and function pointers are used to keep track of the changes to this context.
A simpler approach would be to first gather the information needed for this context (`opVersionName`, `opName`, not sure about `podNames`) then create an instance of this context and not changing it afterwards. This way you'd get rid of a lot of function pointers and simplify the code substantially.

Reply:
I thought about this too. But while simplifying some things, it makes the other more complex: currently, collecting `Instance`, `Operator`, and `OperatorVersion` resources is treated like any other resource and **not** as some sort of a prerequisite (which will require a pre-processing step). So it's homogenous collectors list and dynamic context **vs** extra collecting steps and static context 🤷